src/core/fonts.js

```diff
@@ -482,6 +500,14 @@ function getRanges(glyphs, numGlyphs) {
     }
     codes.push({ fontCharCode: charCode | 0, glyphId: glyphs[charCode] });
   }
+  if (toUnicodeExtraMap) {
+    for (const [unicode, glyphId] of toUnicodeExtraMap) {
+      if (glyphId >= numGlyphs) {
```

Comment:
The following seems to fix the sanitizer errors:
```suggestion
      if (glyphs[unicode] !== undefined || glyphId >= numGlyphs) {
```

Reply:
I fixed it in an other way in adding a check to be sure the unicode is not in the private area:
https://github.com/mozilla/pdf.js/pull/15157/files#diff-d6af99a911b977730586b335a3c7ee702a383cac240414a011ab5534cda1ff3aR476

which should equivalent to what you proposed here.

Reply:
But, that's already present in the latest version of the patch and that failed tests...

Reply:
I don't think so. I just pushed my version:
https://github.com/mozilla/pdf.js/pull/15157/files#diff-d6af99a911b977730586b335a3c7ee702a383cac240414a011ab5534cda1ff3aR477

I think it doesn't really make sense to consider unicode in this area, so it's likely better to just skip them.

Reply:
I suppose that it's *possible*, depending on how the platform font-code handles the `cmap`-table. The fact that both Firefox and Chrome are affected suggests that it's maybe platform, rather than browser, related.

However, I don't use Linux and I also don't know enough about fonts to really tell what happens here unfortunately. Are there any Firefox developers, who are knowledgeable about fonts, that could perhaps help here?

Reply:
> @jfkthame, the reftests on linux (ubuntu) are pretty bad: http://54.241.84.105:8877/7409e99961e9f13/reftest-analyzer.html#web=eq.log
> 
> would you have any idea why changing the cmap could impact the rendering ?

I have a guess.... especially given that this is appearing on Linux, where we use freetype to rasterize fonts. I suspect that including the "real" Unicode values in the cmap is causing the freetype autohinter to process the glyphs (because it makes hinting decisions based on its knowledge of particular Unicode ranges), whereas when we only had PUA codepoints, it made no assumptions about what the glyphs should be.

(If that's correct, the change should presumably be an improvement, in general -- once freetype "knows" which glyphs are all supposed to be uppercase Latin letters, for example, it'll do a better job of harmonizing their heights where possible.)

Reply:
> (If that's correct, the change should presumably be an improvement, in general -- once freetype "knows" which glyphs are all supposed to be uppercase Latin letters, for example, it'll do a better job of harmonizing their heights where possible.)

The problem is that in lots of cases it definitely does *not* look like an improvement...

Reply:
Sometimes I see some improvements, sometimes I don't manage to have an opinion and sometimes it isn't really an improvement.
@Snuffleupagus, could you point out some cases where you consider that the rendering is too bad ?

Reply:
> The following is just a few, quickly spotted, examples:
> 
>     * In `tracemonkey` and `issue5644`, some of the glyphs look "squished" vertically. Particularly noticeable in e.g. the letters "f" and "g" in the title on the front-page.

Hmm.... I haven't looked at many of them, but checking e.g. the front page of `issue5644` (zoomed to 400%), the change in "f" and "g" in the title seems much less significant to me than the overall *improvement* in the body text of the article -- the glyphs look much more even in the "Test" image than the "Reference" (compare the opening words "We report on experience...").

Whether the rendering is better or worse in general, I don't know. (Results may well depend on your local font rendering settings, too -- what hinting level and antialiasing modes are selected.)

Reply:
My feeling is that most of the time we've a little more black pixels and less gray pixels out of the glyph bounding box, hence the glyphs look like better aligned on the base line.

Reply:
I tested a few PDFs (`tracemonkey`, `issue5644` and `issue3115`) locally before/after the patch and I couldn't see any difference at all.

Reply:
> @jfkthame, could you imagine something really bad because of this patch ? I'll test it on mac to see how it behaves.

No, it seems to me like it should be harmless, except for the potential effect -- either good or bad -- on freetype's autohinting of the glyphs (which may or may not happen, depending on the user's font rendering settings).

It'll be interesting to see how the Mac backend behaves with this; does it still remap character codes in the output font in the way I observed in https://bugzilla.mozilla.org/show_bug.cgi?id=1778484#c7, or does it use the "real" Unicode codepoints? I'd hope for the latter, as that will make the resulting PDF much more useful.

Reply:
I just tested on my mac if the patch is correctly fixing the bug 1778484 and yes it's ok.